features/329-principal-apis/api.md

```diff
@@ -0,0 +1,270 @@
+    loginIDKey: string;
+    loginID: string;
+    realm: string;
+}
+interface OAuthIdentity extends IdentityBase {
+    type: 'oauth';
+    providerID: string;
+    providerUserID: string;
+    metadata: {
+        raw_profile: object;
+    };
+}
+interface CustomTokenIdentity extends IdentityBase {
+    type: 'custom';
+    providerUserID: string;
+    metadata: {
+        raw_profile: object;
+    };
+}
+type Identity = PasswordIdentity | OAuthIdentity | CustomTokenIdentity;
+
+async function listIdentities(): Promise<Identity[]>;
+
+async function addLoginID(loginID: {[key: string]: string}, realm?: string): Promise<void>;
+async function removeLoginID(loginID: string, realm?: string): Promise<void>;
+
+async function changePassword(newPassword: string, oldPassword?: string): Promise<User>;
+
+```
+
+
+## Configuration
+`reauthForSecurity` (boolean; default `true`):
+Require re-authentication (i.e. access token must be issued recently) for
+security-critical operations:
+- add/remove login ID
+- link/unlink SSO provider (except auto-link when logging in with OAuth)
+- change password (if old password is not provided)
+
+`reauthIntervalSeconds` (integer; default 300, i.e. 5 minutes):
+The period after the access token is issued, that is considered authenticated
+for security-critical operations.
+
+
+## Use Case Example
+
+### List user identities
+```typescript
+const identities = await listIdentities();
+expectEquals(identities, [
+    {
+        type: "password",
+        loginIDKey: "email",
+        loginID: "test@example.com",
+        realm: "default",
+        metadata: {
+            "email": "test@example.com"
+        }
+    },
+    {
+        type: "password",
+        loginIDKey: "contact_phone",
+        loginID: "+85299999999",
+        realm: "default",
+        metadata: {
+            "phone": "+85299999999"
+        }
+    },
+    {
+        type: "password",
+        loginIDKey: "fingerprint",
+        loginID: "ZmluZ2VycHJpbnQ=",
+        realm: "default",
+        metadata: {}
+    },
+    {
+        type: "oauth",
+        providerID: "some-site",
+        providerUserID: "9999999999999999",
+        metadata: {
+            raw_profile: {
+                id: "9999999999999999",
+                email: "test@example.com",
+                contact_phone: "+85299999999",
+                gender: "male",
+            },
+            email: "test@example.com",
+            phone: "+85299999999",
+        },
+    },
+]);
+```
+
+### Add login ID
+```typescript
+let identities: Identity[];
+identities = await listIdentities();
+expectEquals(identities, [
+    {
+        type: "password",
+        loginIDKey: "email",
+        loginID: "test@example.com",
+        realm: "default",
+        metadata: {
+            "email": "test@example.com"
+        }
+    },
+]);
+
+await addLoginID("secondary_email", "test@oursky.com");
+// throws error "access token is not issued recently"
+
+// after re-authentication:
```

Comment:
Which api the user should use for re-authentication? It might be tricky for users to reconstruct the login API from the current identity, and worth us to provide a shortcut API at client side.

Reply:
Just the normal login API. I expect the flow would be similar to Google's change password: user is prompt to login again when attempting to change password, and user can enter just the new password after logging in.

A dedicated API cannot be provided. User may be logged in using password/OAuth; it is up to app developer to provide suitable UI for re-authentication. App developer can get the way user logged in through current identity.

Reply:
Right, that should be a feature when we start develop a AuthUI Kit.